onnxruntime/core/optimizer/graph_transformer_mgr.h

```diff
@@ -5,31 +5,62 @@
 
 #include "core/optimizer/graph_transformer.h"
 #include "core/optimizer/constant_folding.h"
+#include "core/optimizer/rewrite_rule.h"
+using namespace ::onnxruntime::common;
 
 namespace onnxruntime {
+
+// Older GCC versions don't support std::hash with enum types 
+// Therefore, std::hash<T> appears to be undefined when T is an enum Type. This is fixed in version 6.1
+// TODO: remove this when we update to 6.1 or later
+struct EnumHashKey {
+  template <typename T>
+  uint32_t operator()(T t) const {
+    return static_cast<uint32_t>(t);
+  }
+};
```

Comment:
Either this should be in a general helper header (ideally with checks around whether it's needed or not) or inside the GraphTransformerManager class so it's not at the onnxruntime level.

The latter is quicker, but if you want to do the former it should possibly be defined using std::underlying_type<Enum>::type ([https://en.cppreference.com/w/cpp/utility/hash](https://en.cppreference.com/w/cpp/utility/hash)) as std::underlying_type is available in C++11, and return size_t not uint32_t so it matches the 'real' std::hash better.

Reply:
Or just use uint32_t as the key in the unordered_map instead of TransformerLevel. Usage is internal only, so casting the enum to uint32_t when accessing the map would be an implementation detail.

---
In reply to: [266296860](https://github.com/Microsoft/onnxruntime/pull/608#discussion_r266296860) [](ancestors = 266296860)

Reply:
I was debating between adding to the class or to the namespace... I will move it inside the class... I dont want to change the unordered_map key to uint32 because keeping the enum is more intuitive and straightforward...even though it means the same...plus this is already fixed in gcc 6.1... so when we update removing this hash key will be a trivial change

Reply:
I was debating between adding to the class or to the namespace... I will move it inside the class... I dont want to change the unordered_map key to uint32 because keeping the enum is more intuitive and straightforward...even though it means the same...plus this is already fixed in gcc 6.1... so when we update removing this hash key will be a trivial change